tests/test_functional.py

```diff
@@ -87,6 +90,81 @@ def test_overlay(self, dir):
         compare(config.user, expected=2)
         compare(config.file, expected=3)
 
+    def test_lazy_load(self, dir):
```

Comment:
@RonnyPfannschmidt - okay, try this now. I think formalising the `Proxy` object gives a really nice clean outcome, but interested to hear what you think. If you want, I think you should be fine to stick the proxy object on the config root too :-)

Reply:
from my pov this is basically a drunk identity copied contextvar,

hence the desire to access the config root and path of the dynamic lookup

the tricky point about the client landing in data however is a valid concern, however i would currently consider it less of a problem than having the loader proxy object be integrated differently

how about the concept of a `Sidechannel` value

it would be a `Contant Data Value` that one could update to a object for within a Config, but uppon copying a config, it would
"loose" the Value

then it would no longer be a contextvar, but rather a symblic reference one could map to other objects

Reply:
"from my pov this is basically a drunk identity copied contextvar," comes across quite badly, I'm trying to help here. I don't need this functionality myself, and I'm struggling to understand the resistance to the suggestions I'm making.

I haven't used `ContextVar` so maybe this is a good opportunity for me to learn.

I can't get my head around your `Con(s?)tant Data Value` idea, but it seems to be related to your fixation with storing `vault_loader` on the `Config` object. That's not something I really like, but I'd like to understand why it's important to you so that we can find something that keeps both of us happy :-)

Reply:
i would be fine with not storing the vault loader on the config,
a key consideration for me is how to make it available to the loaded value (which ideally doesn't need a extra magical object

having a proxy object thats tricky to manage and weaseling it into a dynamic loader is something i would like to avoid

im under the impression we have a very different understanding of how to pass the context in (my preference would be to just pass the root config object)

your preference seems to be to pass in something like a context variable, however the proposed variant looks to me like it will carry context over into new copies of the configuration (which seems to be a bug NO)

the idea of the "standin values" would be to have a way to have a sidechannel for the configruation


so that if one had a config like

```py
symbol = Symbol("vault-loader")
config = Config({"password": VaultLoader(loader = symbol, key="password"))
symbol.set_for(config, VaultClient.for_config(config))
assert config.password = "abc123#supersecure"
```

Reply:
first i want to thank you for bringing up how i have been an failure of tone
i apologize, i let unrelated frustration bled into my communication and thus put something on you that you do *not** deserve



first let me outline some details that will hopefully help


1. i believe the get method is not enough for having good data value resolving

i propose adding a `Node.get_context(self, key, root: config, path: configuratior.path.Path)`

this will be able to implement fin resolvers like:

```python
class FormatLaterValue(DataValue):
   def get_context(self, root, path):
      return self.value.format(config=config, this=config.get(path[:-1])

   def get(self):
      raise NotImplementedError("FormatLaterValue requires contet for loading")
```

```yaml
--- defaults.yml
database:
   protocol: postgres
   host: localhost
   username: default
   password: changeme
   port: 9000
   uri = !format "{this.protocol}://{this.username}:{this.password}@{this.host}:{this.port}/myapp" # poor  mans sqla url maker
--- local overrides.yaml

database:
   username: myapp
   password: abc123

```

```python
config = load_and_merge("default.yml", "database.yml") # adds custom tags + loads and merges in order
assert config.database.uri == "postgres://myapp:abc123@localhost:9000/myapp"
```


2. introducing a concept of symbols, which are values that resolve against a configuration

```python

class Symbol(DataValue):
   key: str
   def get_context(self, root, path):
     return self.root.get_symbol(self)
```

when copying a configuration, one would either also copy the "symbol" table (believe this needs a better name)
it it would turn out empty and raise lookup errors

that "symbol/context" table would provide lazy loaders in the configruation with a way to store/retrieeve contextual details


then in turn a vault value would just look up the vaultoader on the root, and use it

as for the issue with carrying over proxies, its not oblivious how to undo/replace them,
take for example the vault loader - if it turns necessary to replace it, i would have to find the proxy objects and replace it, but that in turn would also replace it in the old configuration


please note that the "idea/concept of a context/symbol table is still rather fuzzy, on the dynaconf side, one would simply make a merged config that has fitting root keys and have the resolver "abuse these"

here im trying to avoid just dumping extra objects into the config and rather giving them a defined place to provide context of the configuration to help the resolves find the correct objects

Reply:
What would call `get_context` and where would it get the `root` from?

Reply:
All config nodes up the chain,

Reply:
How about this?

```
        db = Proxy()

        class User(DataValue):

            def __init__(self, db, name):
                self.db = db
                self.name = name

            def get(self):
                db_ = self.db.get()
                return f'{self.name} from {db_}'

        config1 = Config({'users': [User(db, 'a')]})
        config2 = Config({'users': [User(db, 'b')]})

        db.set(0)
        compare(config1.users, expected=['a from 0'])
        compare(config2.users, expected=['b from 0'])

        config1.set_proxy(db, 1)
        config2.set_proxy(db, 2)
        compare(config1.users, expected=['a from 1'])
        compare(config2.users, expected=['b from 2'])
        
        config3 = config1 + config2
        compare(config1.users, expected=['a from 1'])
        compare(config2.users, expected=['b from 2'])
        compare(config3.users, expected=['a from 1', 'b from 2'])

        config3.set_proxy(db, 3)
        compare(config1.users, expected=['a from 1'])
        compare(config2.users, expected=['b from 2'])
        compare(config3.users, expected=['a from 3', 'b from 3'])
```

Reply:
This example also needs a get_context method


An identity behaviour of the proxy would no longer be a concern if there was a get_context on them which would just call get_proxy

Reply:
Please can you show me how you'd use `get_context` in the above example?

Reply:
Most likely the same way as in symbol, which is just a name for proxy

Reply:
`get_context` in your example appears to only be needed if you require path traversal, which I think we both agreed was a bad thing? (The path can be out of sync with where the node is in the config).

I'm trying to figure out a way the `Proxy` can have a value per config without needing a direct reference to the config, since I'm still not sure that can be plumbed through: configurator has always tried to not get involved with the underlying config data (although I think #9 may have to change that...).

Reply:
it cannot be done with `get` - `get_context` is intended to pass the config trough to each constructed node, so it can be used in the lazy resolving

Reply:
What, specifically, cannot be done with `get`?

Reply:
Obtaining the config object that was in use for the lookup